OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+              QueryGpuTimestampNs(device, query_pool, slot_index, timestamp_period);
+          command_buffer_proto->set_begin_gpu_timestamp_ns(begin_timestamp);
+
+          query_slots_to_reset.push_back(slot_index);
+        }
+
+        uint32_t slot_index = completed_command_buffer.command_buffer_end_slot_index;
+        uint64_t end_timestamp =
+            QueryGpuTimestampNs(device, query_pool, slot_index, timestamp_period);
+
+        command_buffer_proto->set_end_gpu_timestamp_ns(end_timestamp);
+        query_slots_to_reset.push_back(slot_index);
+      }
+    }
+  }
+
+  void WriteDebugMarkers(const internal::QueueSubmission& completed_submission,
+                         orbit_grpc_protos::GpuQueueSubmission* submission_proto,
+                         std::vector<uint32_t>& query_slots_to_reset, VkDevice device,
+                         VkQueryPool query_pool, const float timestamp_period) {
+    submission_proto->set_num_begin_markers(completed_submission.num_begin_markers);
+    for (const auto& marker_state : completed_submission.completed_markers) {
+      uint64_t end_timestamp = QueryGpuTimestampNs(
+          device, query_pool, marker_state.end_info->slot_index, timestamp_period);
+      query_slots_to_reset.push_back(marker_state.end_info->slot_index);
+
+      orbit_grpc_protos::GpuDebugMarker* marker_proto = submission_proto->add_completed_markers();
+      if (vulkan_layer_producer_ != nullptr) {
+        marker_proto->set_text_key(
+            vulkan_layer_producer_->InternStringIfNecessaryAndGetKey(marker_state.text));
+      }
+      if (marker_state.color.red != 0.0f || marker_state.color.green != 0.0f ||
+          marker_state.color.blue != 0.0f || marker_state.color.alpha != 0.0f) {
+        auto color = marker_proto->mutable_color();
+        color->set_red(marker_state.color.red);
+        color->set_green(marker_state.color.green);
+        color->set_blue(marker_state.color.blue);
+        color->set_alpha(marker_state.color.alpha);
+      }
+      marker_proto->set_depth(marker_state.depth);
+      marker_proto->set_end_gpu_timestamp_ns(end_timestamp);
+
+      // If we haven't captured the begin marker, we'll leave the optional begin_marker empty.
+      if (!marker_state.begin_info.has_value()) {
+        continue;
+      }
+      orbit_grpc_protos::GpuDebugMarkerBeginInfo* begin_debug_marker_proto =
+          marker_proto->mutable_begin_marker();
+      WriteMetaInfo(marker_state.begin_info->meta_information,
+                    begin_debug_marker_proto->mutable_meta_info());
+
+      uint64_t begin_timestamp = QueryGpuTimestampNs(
+          device, query_pool, marker_state.begin_info->slot_index, timestamp_period);
+      query_slots_to_reset.push_back(marker_state.begin_info->slot_index);
+
+      begin_debug_marker_proto->set_gpu_timestamp_ns(begin_timestamp);
+    }
+  }
+
+  // We use 0 to disable filtering of markers.
+  uint32_t max_local_marker_depth_per_command_buffer_ = 0;
```

Comment:
As commented above, I would consider it clearer if "disable filtering" would be done by setting this to numeric_limits<uint32_t>::max(). That seems to simplify things a bit and you still have the correct interpretation for 0 (depth of 0, that is, no debug labels).